src/server/chromium/crNetworkManager.ts

```diff
@@ -542,3 +563,133 @@ const errorReasons: { [reason: string]: Protocol.Network.ErrorReason } = {
   'timedout': 'TimedOut',
   'failed': 'Failed',
 };
+
+type RequestInfo = {
+  requestId: string,
+  responseReceived: boolean,
+  responseReceivedExtraInfo: Protocol.Network.responseReceivedExtraInfoPayload[],
+  responses: network.Response[],
+  loadingFinished?: Protocol.Network.loadingFinishedPayload,
+  loadingFailed?: Protocol.Network.loadingFailedPayload
+};
+
+//
+// This class aligns responses with response headers from extra info:
+//   - Network.requestWillBeSent, Network.responseReceived, Network.loadingFinished/loadingFailed are
+//     dispatched using one channel.
+//   - Network.requestWillBeSentExtraInfo and Network.responseReceivedExtraInfo are dispatches on
+//     another channel. Those channels are not associated, so events come in random order.
+//
+// This class will associate responses with the new headers. These extra info headers will become
+// available to client reliably upon requestfinished event only.
+//
+class ResponseExtraInfoTracker {
+  private _requests = new Map<string, RequestInfo>();
+
+  constructor() {
+  }
+
+  requestWillBeSent(event: Protocol.Network.requestWillBeSentPayload) {
+    const info = this._requests.get(event.requestId);
+    if (info) {
+      // This is redirect.
+      if (!event.redirectResponse?.connectionId) {
```

Comment:
It seems like both connectionId checks could be done in processResponse method that is called for redirects and final response?

Reply:
We could do that. But a) we would need to collect all the `responseReceived` events for that to be able to check for the connectionId and b) I think there is certain simplicity in following CDP rather than making sense of things as you are reading this code.